### PR TITLE
Live intervals analysis

### PR DESCRIPTION
**Pull request by etiotto**
Add an analysis of a region (e.g. a loop) to determine the live intervals of the values used in the loop.

### CODE REVIEW

**Comment by Dewei-Wang-sh** [comment]
any plan to upstream this?

**Review by chengjunlu — APPROVED** [review]

File: third_party/intel/lib/Analysis/Liveness.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+      OS.indent(2) << LI << "\n";
+  }
+}
+
+LivenessAnalysis::LivenessAnalysis(Operation *op) : mlir::Liveness(op) {
+  assert(getOperation() && !op->getRegions().empty() &&
+         "root operation should not be null and should contain a region");
+  for (Region &rgn : op->getRegions())
+    for (Block &block : rgn)
+      computeLiveIntervals(block);
```
Comment (jopperm) [inline third_party/intel/lib/Analysis/Liveness.cpp:50]: Does this do the right thing for operations actually containing multiple blocks, or would we need proper DFA here? The test pass only covers `scf.for` which has one region and block.

**Comment by etiotto** [comment]
> any plan to upstream this?

Np because there isn't a use case upstream (AFAIK) for using this analysis upstream. I only added it because we have the scheduling pass in our code base (and we might need it for splitting large 2D block loads at some point).

File: third_party/intel/lib/Analysis/Liveness.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+namespace mlir::triton::gpu::intel {
+
+raw_ostream &operator<<(raw_ostream &OS, const LiveInterval &LI) {
+  OpPrintingFlags flags;
+  flags.skipRegions();
+
+  auto printOp = [&](Operation *op) {
+    switch (op->getNumResults()) {
+    case 0:
+      OS << op->getName();
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Analysis/Liveness.cpp:15]: Maybe better to just print the instruction, as name doesn''t uniquely identify an operation. WDYT?

File: third_party/intel/lib/Analysis/Liveness.cpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  const LivenessBlockInfo *livenessInfo = getLiveness(block);
+  if (!livenessInfo)
+    return 0;
+
+  unsigned max = 0;
+  for (Operation &op : block->getOperations()) {
+    unsigned num = numOverlappingLiveIntervals(&op, *livenessInfo);
+    max = (max > num) ? max : num;
+  }
+  return max;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Analysis/Liveness.cpp:73]: should we use `llvm::max_element` here?

File: third_party/intel/lib/Analysis/Liveness.cpp (hunk 2 above)
Comment (etiotto) [reply third_party/intel/lib/Analysis/Liveness.cpp:15]: True, however I find that print the whole instruction is too much text.

File: third_party/intel/lib/Analysis/Liveness.cpp (hunk 3 above)
Comment (etiotto) [reply third_party/intel/lib/Analysis/Liveness.cpp:73]: That one works over ranges, here we just have a scalar.

File: third_party/intel/lib/Analysis/Liveness.cpp (hunk 1 above)
Comment (etiotto) [reply third_party/intel/lib/Analysis/Liveness.cpp:50]: It does compute the live intervals in each block. But you are correct the lit test should also test the case a region contain more than one basic block. I will expand the test case.

**Pull request merged**